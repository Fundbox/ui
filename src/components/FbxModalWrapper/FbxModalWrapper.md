# FbxModalWrapper

#### Note: This component is only the wrapping of a modal to contain it's `header`, `body`, `close-button` and their styles and should be used in conjuction with the vuex modal store.
## Usage

* All modals should implement a `fbx-modal-wrapper` tag.
* `fbx-modal-wrapper` attributes:
    * `title` the modal's title to render. Omitting it will render a modal without the header but only a close button.
    * `close-btn-data-qa` should be unique per modal. This will be the close button data-qa value for testing purposes.
    * `modal-name` is passed as prop into `FbxModalWrapper` so that firing `this.$modal.hide(modalName)` anywhere will have a reference to which modal it should close.
    * `is-loading` will render the loader instead of the slotted children till the loading is `false`.
* `this.$modal.hide` can accept an object as a second argument. This object can be referenced in `before-close` callback handler.
* On desktop viewport sizes A modal has a default width of 600px and it's height set to `auto` to fit it's content.
    * `width` can be overriden by placing a media query on the specific modal implementation scss
* On mobile viewport sizes we enforce the modal to capture the whole screen dimensions.

```html
<fbx-modal-wrapper
  title="BASIC MODAL TITLE"
  close-btn-data-qa="basic-modal-close-btn"
  :modal-name="modalName"
>
  <div v-if="showText">I am shown when showText is true</div>
  <div>some more text</div>
</fbx-modal-wrapper>
```
```html
<fbx-modal-wrapper
  title="LOADING MODAL TITLE"
  close-btn-data-qa="loading-modal-close-btn"
  :modal-name="modalName"
  :is-loading="true"
>
  <div v-if="showText">I am shown when showText is true</div>
  <div>some more text</div>
</fbx-modal-wrapper>

```
```html
<fbx-modal-wrapper
  close-btn-data-qa="no-title-modal-close-btn"
  :modal-name="modalName"
>
  <div v-if="showText">I am shown when showText is true</div>
  <div>some more text</div>
</fbx-modal-wrapper>
```
```html
<fbx-modal-wrapper
  title="SCROLLABLE MODAL TITLE"
  close-btn-data-qa="scrollable-modal-close-btn"
  :modal-name="modalName"
>
  <div>some more text</div>
  <div>some more text</div>
  <div>some more text</div>
  ...
</fbx-modal-wrapper>
```
