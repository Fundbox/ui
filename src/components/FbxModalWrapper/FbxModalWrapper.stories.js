import { storiesOf } from '@storybook/vue'
import { doc } from 'storybook-readme'

import './stories/FbxModalWrapper.stories.scss'
import summary from './FbxModalWrapper.md'
import DemoModalWrapperStoryComponent from './stories/FbxModalWrapper.demo.story'

const stories = storiesOf('Components/ModalWrapper', module)

stories.add('description', doc(summary))
stories.add('demo', () => DemoModalWrapperStoryComponent)

