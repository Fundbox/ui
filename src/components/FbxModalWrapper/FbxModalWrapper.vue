<template>
  <div class="fbx-ui-modal-wrapper layout-column">
    <div :class="[headerClass, 'layout-row', 'vertical-center']">
      <div class="font-18">{{ title }}</div>
      <fbx-close-button
        :data-qa="closeBtnDataQa"
        class="modal-close-button"
        @click="onCloseClick"
      />
    </div>

    <div class="body">
      <div v-if="isLoading" class="loading"></div>
      <div v-else>
        <slot/>
      </div>
    </div>
  </div>
</template>

<script>
import FbxCloseButton from '../FbxCloseButton/FbxCloseButton.vue'
import { isFunction } from 'lodash'

export default {
  components: { FbxCloseButton },
  name: 'FbxModalWrapper',
  destroyed() {
    if (isFunction(this.onClose)) {
      this.onClose()
    }
  },
  props: {
    isLoading: Boolean,
    closeBtnDataQa: String,
    title: String,
    onClose: Function
  },
  computed: {
    headerClass() {
      return this.title ? 'header' : 'no-header'
    }
  },
  inject: ['$validator'],
  methods: {
    onCloseClick() {
      this.$emit('close-btn-clicked')
    }
  }
}
</script>


<style lang="scss">
  @import "../../styles/utils/mixins.scss";
  @import "../../styles/utils/color-palette";

  .v--modal-top-right {
    display: none;
  }

  .v--modal-background-click {
    display: flex;
    align-items: center;
    justify-content: center;
  }

  .v--modal-overlay {
    // $medium-blue with $fbx-backdrop-opacity
    // if we're using the opacity rule then it affects the inner modal box
    background-color: rgba(44, 62, 80, 0.4);
  }

  .v--modal-overlay.scrollable .v--modal-box {
    margin: 0;
  }

  .overlay-transition-enter-active,
  .overlay-transition-leave-active {
    transition: opacity $medium-ease;
  }

  .overlay-transition-enter,
  .overlay-transition-leave-active {
    opacity: 0;
  }

  .box-transition-enter-active,
  .box-transition-leave-active {
    transition: all $medium-bounce;
  }

  .box-transition-enter,
  .box-transition-leave-active {
    opacity: 0;
    transform: translateY(25px);
  }

  .fbx-ui-modal {
    &.v--modal-box {
      position: static;
      @include viewport("sm") {
        margin: 0;
        position: absolute;
        top: 0 !important;
        height: 100% !important;
      }
    }
  }

  .fbx-ui-modal-wrapper {
    position: relative;
    height: 100%;
    background-color: $white;

    .header {
      justify-content: space-between;
      padding: 30px;
      background: $light-gray;
      @include viewport("md") {
        padding: 28px 30px;
      }
    }

    .no-header {
      position: absolute;
      top: 16px;
      right: 16px;
      z-index: 1;
    }

    .modal-close-button {
      font-size: 24px;
      line-height: 24px !important;
    }

    .body {
      padding: 30px;
      position: relative;
      flex: 1;
      @include viewport("sm") {
        overflow-y: auto;
      }
    }

    .loading {
      background-image: url("../../assets/loading-anim.gif");
      background-position: center;
      background-repeat: no-repeat;
      background-color: transparent;
      background-size: 100px;
      display: block;
      top: 0;
      bottom: 0;
      left: 0;
      right: 0;
      content: "";
      position: absolute;
    }
  }

</style>
