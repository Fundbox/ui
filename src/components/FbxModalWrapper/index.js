import Vue from 'vue'
import VModal from 'vue-js-modal'

import FbxModalWrapper from './FbxModalWrapper.vue'
import { registerComponents, vueUse } from '../../utils/plugins'

const components = {
  FbxModalWrapper
}

const VuePlugin = {
  install(Vue) {
    registerComponents(Vue, components)
  }
}

Vue.use(VModal, { componentName: 'fbx-ui-modal' })

vueUse(VuePlugin)

export default VuePlugin
