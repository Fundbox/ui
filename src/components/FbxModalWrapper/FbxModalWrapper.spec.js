import { shallowMount } from '@vue/test-utils'
import FbxModalWrapper from './FbxModalWrapper.vue'

describe('Components/FbxModalWrapper', () => {
  describe('snapshots', () => {
    it('renders the default correctly', () => {
      const wrapper = shallowMount(FbxModalWrapper, {
        // mounting options https://vue-test-utils.vuejs.org/api/options.html
      })
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('renders with loading correctly', () => {
      const wrapper = shallowMount(FbxModalWrapper, {
        propsData: {
          isLoading: true
        }
      })
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('renders without title correctly', () => {
      const wrapper = shallowMount(FbxModalWrapper, {
        propsData: {
          title: undefined
        }
      })
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})
