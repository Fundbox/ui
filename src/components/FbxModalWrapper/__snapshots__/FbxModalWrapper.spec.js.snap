// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/FbxModalWrapper snapshots renders the default correctly 1`] = `
<div class="fbx-ui-modal-wrapper layout-column">
  <div class="header layout-row vertical-center">
    <div class="font-18">snapshot title</div>
    <fbx-close-button-stub class="modal-close-button"></fbx-close-button-stub>
  </div>
  <div class="body">
    <div></div>
  </div>
</div>
`;

exports[`Components/FbxModalWrapper snapshots renders with loading correctly 1`] = `
<div class="fbx-ui-modal-wrapper layout-column">
  <div class="header layout-row vertical-center">
    <div class="font-18">snapshot title</div>
    <fbx-close-button-stub class="modal-close-button"></fbx-close-button-stub>
  </div>
  <div class="body">
    <div class="loading"></div>
  </div>
</div>
`;

exports[`Components/FbxModalWrapper snapshots renders without title correctly 1`] = `
<div class="fbx-ui-modal-wrapper layout-column">
  <div class="no-header layout-row vertical-center">
    <div class="font-18"></div>
    <fbx-close-button-stub class="modal-close-button"></fbx-close-button-stub>
  </div>
  <div class="body">
    <div></div>
  </div>
</div>
`;
